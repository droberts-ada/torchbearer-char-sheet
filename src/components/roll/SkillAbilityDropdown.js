import React from 'react';

const SkillAbilityDropdown = function(props) {
  const value = props.current || "default";
  console.log(`Rendering dropdown with value ${value}`);
  return (
    <select value={value}
      onChange={(event) => {props.onSelectSkill(event.target.value)}}>
      <option key="default" value="default" disabled={true}>
        -- Choose One --
      </option>

      <optgroup label="Abilities">
        {Object.keys(props.character.abilities).map((name) => {
          const ability = props.character.abilities[name];
          return (
            <option
              key={`sd_${name}`}
              value={name}
              >
              {ability.name} ({ability.rating})
            </option>
          );
        })}
      </optgroup>

      <optgroup label="Trained Skills">
      </optgroup>

      <optgroup label="Untrained Skills">
      </optgroup>
    </select>
  );
}

export default SkillAbilityDropdown;
