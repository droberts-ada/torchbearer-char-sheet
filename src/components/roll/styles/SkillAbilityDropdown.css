.skill-ability-dropdown {
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.skill-ability-dropdown li {
  font-size: 1em;
  direction: rtl;
}
.skill-ability-dropdown option {
  direction: ltr;
}
