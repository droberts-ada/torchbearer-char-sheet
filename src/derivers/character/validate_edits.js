import { skillReadyToAdvance } from "../../rules/skills";

// FORMAT
// problems = {
//   wises: {
//     2: {
//      name: ["can't be blank"],
//     }
//   },
//   traits: {
//     1: {
//       name: ["can't be blank"],
//       level: ["must be between 1 and 3"],
//     }
//   }
// }

class ValidationErrors {
  initialize(key, field) {
    if (!this[key]) {
      this[key] = {};
    }

    if (!this[key][field]) {
      this[key][field] = []
    }
  }

  add(key, field, problem) {
    this.initialize(key, field);
    this[key][field].push(problem);
  }
}

const validateNames = (items, errors) => {
  const names = {};

  items.forEach((item, index) => {
    if (!item.name || item.name === '') {
      errors.add(index, 'name', "can't be blank");
    }

    if (names[item.name]) {
      names[item.name].push(index);
    } else {
      names[item.name] = [index];
    }
  });

  Object.keys(names).forEach(name => {
    if (names[name].length > 1) {
      names[name].forEach(index => {
        errors.add(index, 'name', "must be unique");
      });
    }
  });
};

const validateWises = (wises) => {
  const errors = new ValidationErrors();

  validateNames(wises, errors);

  return errors;
};

const validateTraits = (traits) => {
  const errors = new ValidationErrors();

  validateNames(traits, errors);

  traits.forEach((trait, index) => {
    if (trait.level < 1 || trait.level > 3) {
      errors.add(index, 'level', "must be between 1 and 3");
    }
  });

  return errors;
};

const validateRatings = (skills, untaxedNature, errors) => {
  Object.keys(skills).forEach((skillName) => {
    const skill = skills[skillName];
    if (skill.rating > skill.max || skill.rating < skill.min) {
      errors.add(skillName, 'rating', `must be between ${skill.min} and ${skill.max}`);
    }

    // MIGHT doesn't track advancement
    if (skill.advancement) {
      if (skill.advancement.pass < 0) {
        errors.add(skillName, 'pass', "must be greater than 0");
      }

      if (skill.advancement.fail < 0) {
        errors.add(skillName, 'fail', "must be greater than 0");
      }

      if (skillReadyToAdvance(skill, untaxedNature)) {
        errors.add(skillName, 'pass', "skill would advance");
        errors.add(skillName, 'fail', "skill would advance");
      }
    }
  });
};

const validateSkills = (skills, character) => {
  const errors = new ValidationErrors();

  validateRatings(skills, character.abilities.NATURE.untaxed, errors);

  return errors;
}

const validateAbilities = (abilities) => {
  const errors = new ValidationErrors();
  const nature = abilities.NATURE;

  validateRatings(abilities, nature.untaxed, errors);

  if (nature.rating > nature.untaxed) {
    errors.add('NATURE', 'rating', "must be less than or equal to untaxed rating");
  }

  if (nature.untaxed < nature.min || nature.untaxed > nature.max) {
    errors.add('NATURE', 'untaxed', `must be between ${nature.min} and ${nature.max}`);
  }

  return errors;
}

const validateEdits = (character) => {
  if (!character) {
    return undefined;
  }

  return {
    abilities: validateAbilities(character.abilities),
    skills: validateSkills(character.skills, character),
    traits: validateTraits(character.traits),
    wises: validateWises(character.wises),
  };
}

export default validateEdits;