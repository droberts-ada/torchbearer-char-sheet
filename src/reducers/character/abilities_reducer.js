import { MARK_TEST } from '../../actions';
import { advanceSkill } from './skills_reducer';

// See _Abilities_, page 26
const InitialAbilities = {
  NATURE: {
    name: 'Nature',
    rating: 3,
    untaxed: 5,
    min: 0,
    max: 7,
    advancement: {
      pass: 0,
      fail: 0
    },
    open: true, // All abilities are always open
    descriptors: ['Sneaking', 'Riddling', 'Merrymaking']
  },
  WILL: {
    name: 'Will',
    rating: 5,
    min: 1,
    max: 6,
    advancement: {
      pass: 0,
      fail: 0
    },
    open: true
  },
  HEALTH: {
    name: 'Health',
    rating: 3,
    min: 1,
    max: 6,
    advancement: {
      pass: 0,
      fail: 0
    },
    open: true
  },
  RESOURCES: {
    name: 'Resources',
    rating: 3,
    min: 0,
    max: 10,
    advancement: {
      pass: 0,
      fail: 0
    },
    open: true
  },
  CIRCLES: {
    name: 'Circles',
    rating: 4,
    min: 1,
    max: 10,
    advancement: {
      pass: 0,
      fail: 0
    },
    open: true
  },
  MIGHT: {
    name: 'Might',
    rating: 3,
    open: true // do you ever even roll might?
  }
}

const markTest = function (state, action, character) {
  // We use the MARK_TEST action for both skills and abilities
  const ability = action.payload.name;
  if (action.payload.category !== 'abilities' ||
    !state[ability] ||
    ability === 'MIGHT') {
    return state;
  }

  const newState = { ...state };
  newState[ability] = advanceSkill(state[ability], action.payload.mark, character);

  return newState;
}

const abilitiesReducer = function (state = InitialAbilities, action, character) {
  switch (action.type) {
    case MARK_TEST:
      return markTest(state, action, character);

    //
    // case TAX:
    // console.log('TODO: implement tax');
    // return state;
    // break;

    default:
      return state;
  }
}

export default abilitiesReducer;
