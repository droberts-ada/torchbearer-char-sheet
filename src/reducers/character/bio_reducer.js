import { InitialBio } from '../../rules/initial_character';
import {
  RESET_CHARACTER,
  IMPORT_CHARACTER,
  EDIT_CHARACTER_COMMIT
} from '../../actions';

const bioReducer = (state = InitialBio, action) => {
  switch (action.type) {
    case RESET_CHARACTER:
      return InitialBio;

    case IMPORT_CHARACTER:
    case EDIT_CHARACTER_COMMIT:
      return action.payload.bio;

    default:
      break;
  }
  console.log(state);
  return state;
};

export default bioReducer;