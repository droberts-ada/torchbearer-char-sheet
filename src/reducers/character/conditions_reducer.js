import { SET_CONDITION } from '../../actions';
import { Conditions, ConditionRules } from '../../rules/conditions';

const InitialConditions = {};
Object.keys(ConditionRules).forEach((name) => {
  InitialConditions[name] = ConditionRules[name].default_state;
});

const conditionsReducer = function(state=InitialConditions, action) {
  switch (action.type) {
    case SET_CONDITION:
    const name = action.payload.condition;
    const newConditions = {...state};
    newConditions[name] = action.payload.isActive;

    // Getting another condition removes fresh
    if (name !== Conditions.FRESH && newConditions[name]) {
      newConditions[Conditions.FRESH] = false;
    }

    // TODO: can't get fresh until all other conditions are clear
    return newConditions;

    default:
    return state;
  }
}

export default conditionsReducer;
