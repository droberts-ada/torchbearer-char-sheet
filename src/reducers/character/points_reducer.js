import {
  ADD_POINT,
  SPEND_POINT
} from '../../actions';

const InitialPoints = {
  fate: {
    available: 0,
    spent: 0
  },
  persona: {
    available: 0,
    spent: 0
  },
  checks: {
    available: 0,
    spent: 0
  }
};

const pointsReducer = function(state=InitialPoints, action) {
  if (action.type !== ADD_POINT && action.type !== SPEND_POINT) {
    return state;
  }

  const category = action.payload.category;
  const bucket = state[category];
  const newState = {...state};

  switch(action.type) {
    case ADD_POINT:
    newState[category] = {
      ...bucket,
      available: bucket.available + 1
    }
    break;

    case SPEND_POINT:
    if (bucket.available > 0) {
      newState[category] = {
        ...bucket,
        available: bucket.available - 1,
        spent: bucket.spent + 1
      }
    }
    break;

    default:
    throw new Error('How did we even get here?');

  }

  return newState;
};

export default pointsReducer;
