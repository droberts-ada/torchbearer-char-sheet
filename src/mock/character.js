import { ConditionRules } from '../rules/conditions';
import { InitialTraits } from '../reducers/character/traits_reducer';
import { InitialWises } from '../reducers/character/wises_reducer';
import { InitialPoints } from '../reducers/character/points_reducer';

const character = {
  skills: {
    ALCHEMIST: {
      name: 'Alchemist',
      beginnersLuck: 'WILL',
      rating: 0,
      open: false,
      advancement: { pass: 0, fail: 0, },
      min: 0, max: 6,
    },
    FIGHTER: {
      name: 'Fighter',
      beginnersLuck: 'HEALTH',
      rating: 0,
      open: true,
      advancement: { pass: 0, fail: 0, },
      min: 0, max: 6,
    },
    ORATOR: {
      name: 'Orator',
      beginnersLuck: 'WILL',
      rating: 3,
      open: true,
      advancement: { pass: 0, fail: 0, },
      min: 0, max: 6,
    }
  },
  abilities: {
    WILL: {
      name: 'Will',
      rating: 3
    },
    HEALTH: {
      name: 'Health',
      rating: 5
    },
    NATURE: {
      name: 'Nature',
      rating: 4,
      untaxed: 5,
    },
    RESOURCES: {
      name: 'Resources',
      rating: 3
    },
    CIRCLES: {
      name: 'Circles',
      rating: 2
    }
  },
  conditions: {},
  traits: InitialTraits,
  wises: InitialWises,
  points: InitialPoints,
};

Object.keys(ConditionRules).forEach((name) => {
  character.conditions[name] = ConditionRules[name].default_state;
});

// mock character is not fresh (makes things tricky)
character.conditions["FRESH"] = false;

export default Object.freeze(character);
