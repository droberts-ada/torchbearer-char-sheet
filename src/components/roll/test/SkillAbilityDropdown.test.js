import React from 'react';
import renderer from 'react-test-renderer';

import SkillAbilityDropdown from '../SkillAbilityDropdown';

import character from '../../../mock/character';

it('renders with no skill selected', () => {
  const tree = renderer
    .create(
      <SkillAbilityDropdown
        character={character}
        onSelectSkill={skill => skill}
        />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders with a trained skill selected', () => {
  const tree = renderer
    .create(
      <SkillAbilityDropdown
        current="ORATOR"
        character={character}
        onSelectSkill={skill => skill}
        />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
