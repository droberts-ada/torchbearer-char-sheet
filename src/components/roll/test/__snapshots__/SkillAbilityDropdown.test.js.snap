// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with a trained skill selected 1`] = `
<li
  className="skill-ability-dropdown control"
>
  <div>
    <h3>
      Skill / Ability
    </h3>
    <select
      onChange={[Function]}
      value="ORATOR"
    >
      <option
        disabled={true}
        value="default"
      >
        -- Choose One --
      </option>
      <optgroup
        label="Abilities"
      >
        <option
          value="WILL"
        >
          Will
           (
          3
          )
        </option>
        <option
          value="HEALTH"
        >
          Health
           (
          5
          )
        </option>
        <option
          value="NATURE"
        >
          NATURE
           (
          4
          )
        </option>
      </optgroup>
      <optgroup
        label="Trained Skills"
      >
        <option
          value="FIGHTER"
        >
          Fighter
           (
          0
          )
        </option>
        <option
          value="ORATOR"
        >
          Orator
           (
          3
          )
        </option>
      </optgroup>
      <optgroup
        label="Beginner's Luck"
      >
        <option
          value="ALCHEMIST"
        >
          Alchemist
            (
          W
           - 
          3
          )
        </option>
      </optgroup>
    </select>
  </div>
  
</li>
`;

exports[`renders with no skill selected 1`] = `
<li
  className="skill-ability-dropdown control"
>
  <div>
    <h3>
      Skill / Ability
    </h3>
    <select
      onChange={[Function]}
      value="default"
    >
      <option
        disabled={true}
        value="default"
      >
        -- Choose One --
      </option>
      <optgroup
        label="Abilities"
      >
        <option
          value="WILL"
        >
          Will
           (
          3
          )
        </option>
        <option
          value="HEALTH"
        >
          Health
           (
          5
          )
        </option>
        <option
          value="NATURE"
        >
          NATURE
           (
          4
          )
        </option>
      </optgroup>
      <optgroup
        label="Trained Skills"
      >
        <option
          value="FIGHTER"
        >
          Fighter
           (
          0
          )
        </option>
        <option
          value="ORATOR"
        >
          Orator
           (
          3
          )
        </option>
      </optgroup>
      <optgroup
        label="Beginner's Luck"
      >
        <option
          value="ALCHEMIST"
        >
          Alchemist
            (
          W
           - 
          3
          )
        </option>
      </optgroup>
    </select>
  </div>
  
</li>
`;
